http/src/main/java/io/micronaut/http/body/RawMessageBodyHandlerRegistry.java

```diff
@@ -0,0 +1,352 @@
+        HandlerKey<T> key = new HandlerKey<>(type, mediaTypes);
+        MessageBodyReader<?> messageBodyReader = readers.get(key);
+        if (messageBodyReader == null) {
+            MessageBodyReader<T> reader = rawHandler(type, false);
+            if (reader == null) {
+                reader = findReaderImpl(type, mediaTypes);
+            }
+            if (reader != null) {
+                readers.put(key, reader);
+                return Optional.of(reader);
+            } else {
+                readers.put(key, NO_READER);
+                return Optional.empty();
+            }
+        } else if (messageBodyReader instanceof NoReader) {
+            return Optional.empty();
+        } else {
+            //noinspection unchecked
+            return Optional.of((MessageBodyReader<T>) messageBodyReader);
+        }
+    }
+
+    protected abstract <T> MessageBodyWriter<T> findWriterImpl(Argument<T> type, List<MediaType> mediaTypes);
+
+    @SuppressWarnings({"unchecked"})
+    @Override
+    public <T> Optional<MessageBodyWriter<T>> findWriter(Argument<T> type, List<MediaType> mediaType) {
+        HandlerKey<T> key = new HandlerKey<>(type, mediaType);
+        MessageBodyWriter<?> messageBodyWriter = writers.get(key);
+        if (messageBodyWriter == null) {
+            MessageBodyWriter<T> writer = rawHandler(type, true);
+            if (writer == null && type.getType() != Object.class) {
+                writer = findWriterImpl(type, mediaType);
+            }
+            if (writer != null) {
+                writers.put(key, writer);
+                return Optional.of(writer);
+            } else {
+                writers.put(key, NO_WRITER);
+                return Optional.empty();
+            }
+        } else if (messageBodyWriter instanceof NoWriter) {
+            return Optional.empty();
+        } else {
+            //noinspection unchecked
+            return Optional.of((MessageBodyWriter<T>) messageBodyWriter);
+        }
+    }
+
+    private static void addContentType(MutableHeaders outgoingHeaders, @Nullable MediaType mediaType) {
+        if (mediaType != null && !outgoingHeaders.contains(HttpHeaders.CONTENT_TYPE)) {
+            outgoingHeaders.set(HttpHeaders.CONTENT_TYPE, mediaType);
+        }
+    }
+
+    private record RawEntry(Class<?> type, MessageBodyHandler<?> handler) {
+    }
+
+    record HandlerKey<T>(Argument<T> type, List<MediaType> mediaTypes) {
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            HandlerKey<?> that = (HandlerKey<?>) o;
+            return type.equalsType(that.type) && mediaTypes.equals(that.mediaTypes);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(type.typeHashCode(), mediaTypes);
```

Comment:
You can use `io.micronaut.core.util.ObjectUtils` that avoids new array allocation. You might want to store the hash code.

Reply:
will use ObjectUtils. storing the hash shouldn't matter, the map caches it anyway.